modules/firehose/module.go

```diff
@@ -228,33 +237,69 @@ const configSchemaString = `
 `
 
 type Module struct {
-	schema *gjs.Schema
+	schema             *gjs.Schema
+	providerRepository *mongodb.ProviderRepository
 }
 
 func (m *Module) ID() string {
 	return "firehose"
 }
 
-func New() *Module {
+func New(providerRepository *mongodb.ProviderRepository) *Module {
 	schemaLoader := gjs.NewStringLoader(configSchemaString)
 	schema, err := gjs.NewSchema(schemaLoader)
 	if err != nil {
 		return nil
 	}
 	return &Module{
-		schema: schema,
+		schema:             schema,
+		providerRepository: providerRepository,
 	}
 }
 
 func (m *Module) Apply(r *domain.Resource) (domain.ResourceStatus, error) {
+	for _, p := range r.Providers {
+		provider, err := m.providerRepository.GetByURN(p.Urn)
+		if err != nil {
+			return domain.ResourceStatusError, err
+		}
+
+		if provider.Kind == KUBERNETES {
+			releaseConfig := helm.DefaultReleaseConfig()
+			var values = make(map[string]interface{})
+
+			kubeConfig := helm.ToKubeConfig(provider.Configs)
+			helmConfig := &helm.ProviderConfig{
+				Kubernetes: kubeConfig,
+			}
+			helmProvider := helm.NewProvider(helmConfig)
+
+			v := reflect.ValueOf(r.Configs[valuesConfigString])
+			if err := mapstructure.Decode(v, &values); err != nil {
+				return domain.ResourceStatusError, err
+			}
+			releaseConfig.Values = values
+
+			err := mapstructure.Decode(r.Configs, &releaseConfig)
+			if err != nil {
+				return domain.ResourceStatusError, err
+			}
+
+			_, err = helmProvider.Release(releaseConfig)
+			if err != nil {
+				return domain.ResourceStatusError, nil
+			}
+		}
+	}
+
 	return domain.ResourceStatusCompleted, nil
 }
 
 func (m *Module) Validate(r *domain.Resource) error {
-	resourceLoader := gjs.NewGoLoader(r.Configs)
+	resourceLoader := gjs.NewGoLoader(r.Configs[valuesConfigString])
```

Comment:
We should also validate the other fields of `r.Configs` apart from `r.Configs[valuesConfigString]`.

Reply:
Currently, the firehose module's `Validate` method is only validating, firehose-related configurations.

The other configuration, that the user sends are stored directly into DB, and in the `Apply` method, it is `merged` with the default configs, before release.

So, `where` shall we validate the other configurations?

Reply:
We can follow either of the following two approaches,
- Prepare `configSchemaString` for the helm, siren, etc providers, as we add them, and validate them inside the `Validate` method using JSON Schema validator.
- Add validation checks, and default values in the struct definition itself  of each providers. And, use a [validator](https://github.com/go-playground/validator), to validate, before creating a resource.

Both of these approaches would require us to validate `resource.Configs` against all the `kind` of providers, passed with a `resource`.

Reply:
I think we need to basically treat these configs as a part of the firehose configs itself. Instead of treating them as external. We should also figure out a way to compute most of them from other config fields.